Write errors are not returned from WriteTo for short compressed files

It seems that write errors from `Decoder.WriteTo` are not returned when the compressed input is short. This causes `io.Copy` calls to wrongly return success, even though a write has failed. See the test case below for an example. 

```golang
type failingWriter struct {
	err error
}

func (f failingWriter) Write(_ []byte) (n int, err error) {
	return 0, f.err
}

func TestErrorWriter(t *testing.T) {
	input := make([]byte, 100)
	cmp := bytes.Buffer{}
	w, err := NewWriter(&cmp)
	if err != nil {
		t.Fatal(err)
	}
	defer w.Close()
	_, _ = rand.Read(input)
	_, err = w.Write(input)
	if err != nil {
		t.Fatal(err)
	}
	wantErr := fmt.Errorf("i'm a failure")
	zr, err := NewReader(&cmp)
	defer zr.Close()
	out := failingWriter{err: wantErr}
	_, err = zr.WriteTo(out)
	if !errors.Is(err, wantErr) {
		t.Errorf("error: wanted: %v, got: %v", wantErr, err)
	}
}
```

Test output:
```
=== RUN   TestErrorWriter
    decoder_test.go:195: error: wanted: i'm a failure, got: <nil>
--- FAIL: TestErrorWriter (0.00s)
```

--- a/zstd/decoder.go
+++ b/zstd/decoder.go
@@ -260,9 +260,10 @@ func (d *Decoder) WriteTo(w io.Writer) (int64, error) {
 		if len(d.current.b) > 0 {
 			n2, err2 := w.Write(d.current.b)
 			n += int64(n2)
-			if err2 != nil && d.current.err == nil {
+			if err2 != nil && (d.current.err == nil || d.current.err == io.EOF) {
 				d.current.err = err2
-				break
+			} else if n2 != len(d.current.b) {
+				d.current.err = io.ErrShortWrite
 			}
 		}
 		if d.current.err != nil {

--- a/zstd/decoder_test.go
+++ b/zstd/decoder_test.go
@@ -165,6 +165,40 @@ func TestErrorReader(t *testing.T) {
 	}
 }
 
+type failingWriter struct {
+	err error
+}
+
+func (f failingWriter) Write(_ []byte) (n int, err error) {
+	return 0, f.err
+}
+
+func TestErrorWriter(t *testing.T) {
+	input := make([]byte, 100)
+	cmp := bytes.Buffer{}
+	w, err := NewWriter(&cmp)
+	if err != nil {
+		t.Fatal(err)
+	}
+	_, _ = rand.Read(input)
+	_, err = w.Write(input)
+	if err != nil {
+		t.Fatal(err)
+	}
+	err = w.Close()
+	if err != nil {
+		t.Fatal(err)
+	}
+	wantErr := fmt.Errorf("i'm a failure")
+	zr, err := NewReader(&cmp)
+	defer zr.Close()
+	out := failingWriter{err: wantErr}
+	_, err = zr.WriteTo(out)
+	if !errors.Is(err, wantErr) {
+		t.Errorf("error: wanted: %v, got: %v", wantErr, err)
+	}
+}
+
 func TestNewDecoder(t *testing.T) {
 	defer timeout(60 * time.Second)()
 	testDecoderFile(t, "testdata/decoder.zip")

